### PR TITLE
mgr/orch: new cli, phase 2

### PR DESCRIPTION
**Pull request by liewegas**
Lots more cleanup and simplification of all of the add/update methods, but this is progress.

### CODE REVIEW

File: src/pybind/mgr/volumes/fs/fs_util.py [hunk 1]
```diff
@@ -36,7 +36,7 @@ def remove_filesystem(mgr, fs_name):
 def create_mds(mgr, fs_name):
     spec = orchestrator.ServiceSpec('mds', fs_name)
     try:
-        completion = mgr.add_mds(spec)
+        completion = mgr.service_aply(spec)
```
Comment (tchaikov) [inline src/pybind/mgr/volumes/fs/fs_util.py]:
```suggestion
        completion = mgr. service_apply(spec)
```

**Comment by liewegas** [comment]
passing cephadm tests... http://pulpito.ceph.com/sage-2020-02-13_02:36:46-rados-wip-sage4-testing-2020-02-12-1611-distro-basic-smithi/

File: src/pybind/mgr/orchestrator.py [hunk 2]
```diff
@@ -1339,7 +1340,7 @@ def validate_add(self):
 
 class NFSServiceSpec(ServiceSpec):
     def __init__(self, name, pool=None, namespace=None, placement=None):
-        super(NFSServiceSpec, self).__init__(name, placement)
+        super(NFSServiceSpec, self).__init__('nfs', name, placement)
 
         #: RADOS pool where NFS client recovery data is stored.
         self.pool = pool
```
Comment (sebastian-philipp) [inline src/pybind/mgr/orchestrator.py]: Hm, this makes it hard to deserialize `NFSServiceSpec`s. 

Let's have a yaml

```yaml
service_type: nfs
name: mynfs
pool: mypool
namespace: mynamespace
```

how can we write a function

```python
class ServiceSpec:
    @classmethod
    from_yaml(input: str): ...
```
that returns a `NFSServiceSpec`? I'm a bit against creating a `ServiceSpecFactory`

File: src/pybind/mgr/cephadm/module.py [hunk 3]
```diff
@@ -1488,6 +1488,19 @@ def _filter(daemons):
...
+            return self.update_mgrs(spec)
+        if spec.service_type == 'mon':
+            return self.update_mons(spec)
+        if spec.service_type == 'mds':
+            return self.update_mds(spec)
+        if spec.service_type == 'rgw':
+            return self.update_rgw(spec)
+        if spec.service_type == 'rbd-mirror':
+            return self.update_rbd_mirror(spec)
+        raise NotImplementedError()
```
Comment (sebastian-philipp) [inline src/pybind/mgr/cephadm/module.py]: Can we move this to the orchestrator_cli and keep the Orchestrator interface clean?

File: src/pybind/mgr/rook/module.py [hunk 4]
```diff
@@ -340,6 +340,16 @@ def remove_service(self, service_type, service_name):
...
+    def service_apply(self, spec):
+        if spec.service_type == 'mon':
+            return self.update_mons(spec)
+        if spec.service_type == 'mgr':
+            raise NotImplementedError()
+        if spec.service_type == 'mds':
+            return self.update_mds(spec)
+        if spec.service_type == 'nfs':
+            return self.update_nfs(spec)
+
```
Comment (sebastian-philipp) [inline src/pybind/mgr/rook/module.py]: yep. let's move this to the orch_cli

File: src/pybind/mgr/orchestrator_cli/module.py [hunk 5]
```diff
@@ -179,7 +179,7 @@ def _remove_host(self, host):
         return HandleCommandResult(stdout=completion.result_str())
 
     @orchestrator._cli_write_command(
-        'orchestrator host set-addr',
+        'orch host set-addr',
```
Comment (sebastian-philipp) [inline src/pybind/mgr/orchestrator_cli/module.py]:
```suggestion
        'orch apply host ',
```
?

**Comment by sebastian-philipp** [comment]
doc fails with:

```
Traceback (most recent call last):
  File "/home/jenkins-build/build/workspace/ceph-pr-docs/build-doc/virtualenv/lib/python3.5/site-packages/sphinx/util/inspect.py", line 230, in safe_getattr
    return getattr(obj, name, *defargs)
AttributeError: type object 'Orchestrator' has no attribute 'update_mds'

During handling of the above exception, another exception occurred:
```

File: src/pybind/mgr/orchestrator.py [hunk 6]
```diff
@@ -1305,9 +1305,8 @@ class ServiceSpec(object):
 
     """
 
-    def __init__(self, service_type, name=None, placement=None):
```
Comment (sebastian-philipp) [inline src/pybind/mgr/orchestrator.py]: Sorry for the misunderstanding here! Having a `service_type` within `ServiceSpec` does make sense. As it makes it possible to just do `ceph orch import -i my_cluster.yaml` at some point, which is awesome! But it loosk like as if we need to rethink the class structure

File: src/pybind/mgr/orchestrator.py [hunk 7]
```diff
@@ -878,8 +878,8 @@ def list_daemons(self, daemon_type=None, daemon_id=None, host=None, refresh=Fals
         """
         raise NotImplementedError()
 
-    def remove_daemons(self, names):
-        # type: (List[str]) -> Completion
+    def remove_daemons(self, names, force):
```
Comment (sebastian-philipp) [inline src/pybind/mgr/orchestrator.py]:
```suggestion
    def remove_daemons(self, names, force=False):
```

But we can do this in a different PR.

**Comment by liewegas** [comment]
retest this plesae

**Comment by liewegas** [comment]
retest this please
